jmix-flowui/flowui/src/main/java/io/jmix/flowui/component/navigationmenubar/NavigationMenuBar.java

```diff
@@ -0,0 +1,830 @@
+        if (Objects.equals(this.menuItemProvider, menuItemProvider)) {
+            return;
+        }
+        if (itemCollectionChangedSubscription != null) {
+            itemCollectionChangedSubscription.remove();
+            itemCollectionChangedSubscription = null;
+        }
+        this.menuItemProvider = menuItemProvider;
+        if (menuItemProvider != null) {
+            itemCollectionChangedSubscription =
+                    menuItemProvider.addCollectionChangedListener(this::onMenuItemCollectionChanged);
+        }
+    }
+
+    protected void onMenuItemCollectionChanged(MenuItemProvider.CollectionChangeEvent<AbstractMenuItem<?>> e) {
+        removeAllMenuItems();
+        e.getItems().forEach(this::addMenuItem);
+    }
+
+    /**
+     * Returns menu item from the menu (including non-root items) by its id.
+     *
+     * @param id menu item id
+     * @return menu item or null if not found
+     */
+    @Nullable
+    public AbstractMenuItem<?> getMenuItem(String id) {
+        Preconditions.checkNotNullArgument(id);
+
+        return allMenuItems.get(id);
+    }
+
+    /**
+     * @return immutable list of root menu items
+     */
+    public List<AbstractMenuItem<?>> getMenuItems() {
+        return Collections.unmodifiableList(rootMenuItems);
+    }
+
+    /**
+     * Adds menu item and its children to menu root.
+     *
+     * @param menuItem menu item to add
+     */
+    public void addMenuItem(AbstractMenuItem<?> menuItem) {
+        Preconditions.checkNotNullArgument(menuItem);
+
+        addMenuItemRecursive(menuItem, null, null);
+
+        rootMenuItems.add(menuItem);
+    }
+
+    protected void addMenuItemRecursive(AbstractMenuItem<?> childMenuItem,
+                                        @Nullable ParentMenuItem parentMenuItem,
+                                        @Nullable Integer index) {
+        checkItemIdDuplicate(getMenuItemIdNN(childMenuItem));
+
+        JmixSubMenu parentSubMenu = getParentSubMenu(parentMenuItem);
+
+        if (childMenuItem.isMenu()) {
+            addMenuItemToSubMenu(childMenuItem, parentSubMenu, index);
+
+            ParentMenuItem childParentMenuItem = (ParentMenuItem) childMenuItem;
+            for (AbstractMenuItem<?> item : childParentMenuItem.getChildItems()) {
+                addMenuItemRecursive(item, childParentMenuItem, null);
+            }
+        } else if (childMenuItem.isSeparator()) {
+            if (parentSubMenu != null) {
+                addComponentToSubMenu(childMenuItem, parentSubMenu, index);
+            } else {
+                throw new UnsupportedOperationException("Separators are not supported on the root level");
+            }
+        } else {
+            addMenuItemToSubMenu(childMenuItem, parentSubMenu, index);
+        }
+
+        attachMenuItem(childMenuItem);
+    }
+
+    protected String getMenuItemIdNN(AbstractMenuItem<?> menuItem) {
```

Comment:
You have neither `@Nullable` annotation not pair method, so no need in `NN`, which is considered as legacy naming.